crates/sui-framework/packages/move-stdlib/sources/vector.move

```diff
@@ -370,3 +404,54 @@ public macro fun zip_map_ref<$T1, $T2, $U>(
     zip_do_ref!($v1, $v2, |el1, el2| r.push_back($f(el1, el2)));
     r
 }
+
+/// Iterate through the vector and find the indices of elements that satisfy the predicate `p`.
+/// Returns a new vector containing the indices of elements that satisfy the predicate in the
+/// order they appear in the original vector.
+public macro fun find_indices<$T>($v: &vector<$T>, $p: |&$T| -> bool): vector<u64> {
+    let v = $v;
+    let mut r = vector[];
+    v.length().do!(|i| if ($p(&v[i])) r.push_back(i));
+    r
+}
+
+/// Takes the first `n` elements of the vector `v` that satisfy the predicate `p`.
+/// Modifies the original vector.
+public macro fun take_mut_while<$T>($v: &mut vector<$T>, $p: |&$T| -> bool): vector<$T> {
+    let v = $v;
+    let len = v.length();
+    'search: {
+        len.do!(|i| if (!$p(&v[i])) return 'search v.take_mut(i));
+        let mut r = tabulate!(len, |_| v.pop_back());
+        r.reverse();
+        r
+    }
+}
+
+/// Takes the first `n` elements of the vector `v` that satisfy the predicate `p` and drops the rest.
+/// Destroys the original vector after taking the elements.
+public macro fun take_while<$T: drop>($v: vector<$T>, $p: |&$T| -> bool): vector<$T> {
+    let v = $v;
+    'search: {
+        v.length().do!(|i| if (!$p(&v[i])) return 'search v.take(i));
+        v
+    }
+}
+
+/// Take all elements of the vector `v` except the first `n` elements that satisfy the predicate `p`.
+/// Modifies the original vector.
+public macro fun skip_mut_while<$T>($v: &mut vector<$T>, $p: |&$T| -> bool): vector<$T> {
+    let v = $v;
+    // find first element that does not satisfy the predicate, then skip from there
+    'search: {
+        v.length().do!(|i| if (!$p(&v[i])) return 'search v.skip_mut(i));
+        vector[]
```

Comment:
```suggestion
       let mut r = tabulate!(len, |_| v.pop_back());
       r.reverse();
       r
```

Reply:
I'm not sure if that's correct. If all elements are skipped, you should return an empty vector.

Reply:
This behaviour is opposite to the one in take - in take you take while or return the whole vector.